sunscreen/src/types/zkp/gadgets/arithmetic.rs

```diff
@@ -0,0 +1,221 @@
+use crypto_bigint::UInt;
+use subtle::{ConditionallySelectable, ConstantTimeEq, ConstantTimeGreater};
+use sunscreen_zkp_backend::{BigInt, Gadget};
+
+use crate::{invoke_gadget, with_zkp_ctx, zkp::ZkpContextOps, ZkpResult};
+
+use super::ToUInt;
+
+pub struct SignedModulus {
+    field_modulus: BigInt,
+    max_remainder_bits: usize,
+}
+
+impl SignedModulus {
+    /**
+     * Creates a new Signed modulus gadget.
+     *
+     * # Remarks
+     * To produce efficient circuits, you must specify the maximum number
+     * of bits the remainder requires. This value should be
+     * `ceil(log2(abs(m)))`.
+     * Additionally, this value must be less than `log2(f)` where `f`
+     * is the size of the backend field.
+     *
+     * # Panics
+     * * When `field_modulus == 0`
+     */
+    pub fn new(field_modulus: BigInt, max_remainder_bits: usize) -> Self {
+        assert_ne!(field_modulus, BigInt::ZERO);
+
+        Self {
+            field_modulus,
+            max_remainder_bits,
+        }
+    }
+}
+
+impl Gadget for SignedModulus {
+    fn gen_circuit(
+        &self,
+        gadget_inputs: &[petgraph::stable_graph::NodeIndex],
+        hidden_inputs: &[petgraph::stable_graph::NodeIndex],
+    ) -> Vec<petgraph::stable_graph::NodeIndex> {
+        let q = hidden_inputs[0];
+        let r = hidden_inputs[1];
+        let x = gadget_inputs[0];
+        let m = gadget_inputs[1];
+
+        let m_min_1_min_r = with_zkp_ctx(|ctx| {
+            // assert mq + r == x
+            let t = ctx.add_multiplication(m, q);
+            let res = ctx.add_addition(t, r);
+            let diff = ctx.add_subtraction(x, res);
+
+            ctx.add_constraint(diff, &BigInt::ZERO);
+
+            // Compute m - 1 - r so we can later attempt a binary expansion
+            let one = ctx.add_constant(&BigInt::ONE);
+            let m_min_1 = ctx.add_subtraction(m, one);
+            ctx.add_subtraction(m_min_1, r)
+        });
+
+        // Get a bound [0, 2^k]. This establishes r >= 0.
+        invoke_gadget(ToUInt::new(self.max_remainder_bits), &[r]);
+
+        // Show r < m
+        invoke_gadget(ToUInt::new(self.max_remainder_bits), &[m_min_1_min_r]);
+
+        vec![hidden_inputs[0], hidden_inputs[1]]
+    }
+
+    fn hidden_input_count(&self) -> usize {
+        2
+    }
+
+    fn gadget_input_count(&self) -> usize {
+        2
+    }
+
+    fn compute_inputs(
+        &self,
+        gadget_inputs: &[sunscreen_zkp_backend::BigInt],
+    ) -> ZkpResult<Vec<sunscreen_zkp_backend::BigInt>> {
+        let x = gadget_inputs[0];
+        let m = gadget_inputs[1];
+
+        // As per docs, when shift amount is constant, time is constant.
+        // See https://docs.rs/crypto-bigint/latest/crypto_bigint/struct.UInt.html#method.shr_vartime
+        let fm_2 = self.field_modulus.shr_vartime(2);
+
+        let is_neg = x.ct_gt(&fm_2);
+        let (q_pos, r_pos) = x.div_rem(&m).unwrap();
```

Comment:
So this will panic iff `m == 0`. I don't see us asserting that is the case anywhere, though. Should we add that? Ditto on the `reduce` unwraps below.

If this isn't assertable anywhere, would it make more sense to have this bubble up to a ZkpResult instead of panicking?

Reply:
I actually did have a panic in previous iterations where m was passed in the gadget constructor. I then changed it to just receive `m` as an argument and forgot to add the error condition.

Reply:
Good catch either way.